cluster.go

```diff
@@ -1056,16 +1066,69 @@ func (c *Cluster) pin(pin api.Pin, blacklist []peer.ID, prioritylist []peer.ID)
 // of underlying IPFS daemon unpinning operations.
 func (c *Cluster) Unpin(h *cid.Cid) error {
 	logger.Info("IPFS cluster unpinning:", h)
+	cState, err := c.consensus.State()
+	if err != nil {
+		return err
+	}
 
-	pin := api.Pin{
-		Cid: h,
+	if !cState.Has(h) {
+		return errors.New("cannot unpin pin uncommitted to state")
+	}
+	pin := cState.Get(h)
+
+	switch pin.Type {
+	case api.DataType:
+		return c.consensus.LogUnpin(pin)
+	case api.ShardType:
+		warn := "unpinning shard cid %s without unpinning parent"
+		logger.Warningf(warn, h.String())
+		return c.consensus.LogUnpin(pin)
```

Comment:
Not sure if it should be allowed at all

Reply:
Eventually we will likely want to provide a way for users to evict shards and clusterdags manually from the state at least in the case a crash has occurred and the state should be cleaned.  

Right now given our assumption that data dags are being streamed and cannot fit entirely on a peer during sharding, shard components must be pinned without having a metaPin reference linked to them when shards are complete but the whole clusterdag is not.  We could also avoid this by just block putting everything and waiting to Pin until everything is ready.  Of course all of this belongs in a future discussion around data safety and recovery.  For now I am fine with just erroring when this happens and using `state cleanup` if unlinked shard pins make it into the state.